### DISCUSSION TITLE
I2C driver issue

### DISCUSSION THREAD

**Discussion by andreas-buerkler-enclustra** [polarfire-soc/polarfire-soc-documentation#145] [Linux drivers for PolarFire Soc] [answered]
Hi

When reading more than one byte from I2C bus, I see that too much data is read. This causes issues with some I2C device drivers (e.g. atsha204a).

"i2ctransfer -y 0 w1@0x6f 0 r1" works as expected, only one byte is read.

"i2ctransfer -y 0 w1@0x6f 0 r2" reads 3 bytes instead of two. I2ctransfer returns only 2 bytes but on the I2C bus, 3 bytes are read.

In the driver source code, the CTRL_AA flag is cleared when lenght is zero, which causes an additional byte read with NACK. Changing that condition to one seems to solve the issue. I'm not 100% sure if my patch is correct but it would be great if you can check that.

```diff
diff --git a/drivers/i2c/busses/i2c-microchip.c b/drivers/i2c/busses/i2c-microchip.c
index 3eeda8403b0b..0c7ff5bf35d8 100755
--- a/drivers/i2c/busses/i2c-microchip-core.c
+++ b/drivers/i2c/busses/i2c-microchip-core.c
@@ -207,7 +207,7 @@ static void mchp_corei2c_empty_rx(struct mchp_corei2c_dev *idev)
 		idev->msg_len--;
 	}
 
-	if (idev->msg_len == 0) {
+	if (idev->msg_len <= 1) {
 		ctrl = readb(idev->base + CORE_I2C_CTRL);
 		ctrl &= ~CTRL_AA;
 		writeb(ctrl, idev->base + CORE_I2C_CTRL);
-- 
```

**Comment by ConchuOD** [comment]
Hey,
I see your driver is called "i2c-microchip.c" - but in both the (deprecated) 5.12 kernel and (currently supported) 5.15 kernel, the driver is called "i2c-microchip-core.c".
Could you try updating to a more recent kernel and checking if it is fixed?
Thanks,
Conor.

**Reply by andreas-buerkler-enclustra** [reply]
Hi, I'm currently using 5.12 kernel (I think it is 2021.11 release) because I already have a few kernel patches for my hardware that I will update later to a newer version. The 5.12 I2C driver was not really working. Therefore, I copied the latest version of i2c-microchip-core.c from 5.15 to 5.12 and renamed it to keep the changes simple. The updated driver works much better except the described issue.

**Reply by ConchuOD** [reply]
>The updated driver works much better except the described issue.

Do you mean, "the issue is still present"?

**Reply by andreas-buerkler-enclustra** [reply]
Yes the issue is still present in the latest version of the I2C driver.

**Reply by ConchuOD** [reply]
Thanks for confirming, I'll take a look :)

**Reply by ConchuOD** [reply]
It would appear that the bare-metal driver agrees with you [0]. Your change looks correct to me - I checked with a protocol analsyer and there is indeed an extra transfer.

Is there a name & email address I can attribute a Reported-by: to?

0 - https://github.com/polarfire-soc/platform/blob/main/drivers/mss/mss_i2c/mss_i2c.c#L838

**Reply by andreas-buerkler-enclustra** [reply]
Thank you for having a look at the issue. You can use andreas.buerkler@enclustra.com.

**Reply by ConchuOD** [reply]
> Thank you for having a look at the issue. You can use [andreas.buerkler@enclustra.com](mailto:andreas.buerkler@enclustra.com).

Great, thanks!

**Reply by ConchuOD** [reply]
Not sure if you were notified, but the patch was applied upstream.
Should be mirrored this week or early next to: https://github.com/linux4microchip/linux

Thanks for reporting the issue, greatly appreciated.

**Reply by ConchuOD** [reply]
This is now merged to our git tree also:
https://github.com/linux4microchip/linux/commit/21a3ec9d24565014edc74ae47c008a6bd5d979c3

Thanks for your report and fix!

**Comment by ConchuOD** [comment, marked as answer]
Fixed upstream and in our tree as:
https://github.com/linux4microchip/linux/commit/21a3ec9d24565014edc74ae47c008a6bd5d979c3
